### PR TITLE
pf serving support distributed tracing and feedback api

### PR DESCRIPTION
**Pull request by wxpjimmy**
# Description
changes include:
- support distributed trace context/baggage propagation
- add new feedback colleciton API
- refine code for metrics/trace exporter and add app insight trace support

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_sdk/_serving/utils.py [hunk 1]
```diff
@@ -137,3 +142,48 @@ def encode_dict(data: dict) -> str:
...
+        b2 = {'baggage': baggage}
+        baggage_ctx = W3CBaggagePropagator().extract(b2, context=ctx)
+    else:
+        baggage_ctx = ctx
+    if baggage_ctx:
+        logger.info(f"Received trace context: {baggage_ctx}")
+    return baggage_ctx
+
+
+def serialize_attribute_value(v):
```
Comment (zzn2) [inline src/promptflow/promptflow/_sdk/_serving/utils.py:170]: Do we already have a similar method in the `promptflow.tracing` package? cc @lumoslnt @thy09

File: src/promptflow/promptflow/tracing/contracts/trace.py [hunk 2]
```diff
@@ -16,6 +16,7 @@ class TraceType(str, Enum):
     LANGCHAIN = "LangChain"
     FLOW = "Flow"
     EMBEDDING = "Embedding"
+    FEEDBACK = "Feedback"
```
Comment (zzn2) [inline src/promptflow/promptflow/tracing/contracts/trace.py:19]: Is it good to add a new "Feedback" type here? Is it too specific that only used by one scenario? cc @thy09

File: src/promptflow/promptflow/_sdk/_serving/extension/extension_type.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# ---------------------------------------------------------
+
+from enum import Enum
+
+
+class ExtensionType(Enum):
+    """Extension type used to identify which extension to load in serving app."""
+
+    Default = "local"
```
Comment (zzn2) [inline src/promptflow/promptflow/_sdk/_serving/extension/extension_type.py:11]: Should enum use capitalized names? `Default` -> `DEFAULT`

File: src/promptflow/promptflow/_sdk/_serving/utils.py [hunk 4]
```diff
@@ -137,3 +142,48 @@ def encode_dict(data: dict) -> str:
     b64_data = base64.b64encode(zipped_data)
     # bytes -> str
     return b64_data.decode()
+
+
+def try_extract_trace_context(logger) -> Context:
```
Comment (zzn2) [inline src/promptflow/promptflow/_sdk/_serving/utils.py:147]: It seems that the open telemetry SDK has util functions that do the similar thing:
image

File: src/promptflow/promptflow/_sdk/_serving/app.py [hunk 5]
```diff
@@ -211,6 +204,38 @@ def version():
...
+                        data = json.loads(data)
+                        for k in data:
+                            span.set_attribute(k, serialize_attribute_value(data[k]))
+                    except Exception as e:
+                        logger.warning(f"Failed to flatten the feedback, fall back to non-flattern mode. Error: {e}.")
+                        span.set_attribute('feedback', str(data))
+                else:
+                    span.set_attribute('feedback', data)
+                # add baggage data if exist
+                data = baggage.get_all()
```
[the remark this answers is not in the document]
Comment (wxpjimmy) [reply src/promptflow/promptflow/_sdk/_serving/app.py:230]: customer decide which data to pass, it's optional, this is Otel standard concept.

File: src/promptflow/promptflow/_sdk/_serving/app.py [hunk 6]
```diff
@@ -211,6 +204,38 @@ def version():
...
+                if data:
+                    for k, v in data.items():
+                        span.set_attribute(k, v)
```
[the remark this answers is not in the document]
Comment (wxpjimmy) [reply src/promptflow/promptflow/_sdk/_serving/app.py:233]: No, this is collect what customer passed in,  the field name is also defined by customer, we don't need to add any prefix for this.

File: src/promptflow/promptflow/_sdk/_serving/app.py [hunk 7]
```diff
@@ -211,6 +204,38 @@ def version():
...
+            with open_telemetry_tracer.start_as_current_span('promptflow-feedback') as span:
+                span.set_attribute("span_type", TraceType.FEEDBACK.value)
+                data = request.get_data()
+                should_flattern = request.args.get('flatten', 'false').lower() == 'true'
+                if should_flattern:
+                    try:
+                        # try flatten the data to avoid data too big issue (especially for app insights scenario)
+                        data = json.loads(data)
+                        for k in data:
+                            span.set_attribute(k, serialize_attribute_value(data[k]))
```
[the remark this answers is not in the document]
Comment (wxpjimmy) [reply src/promptflow/promptflow/_sdk/_serving/app.py:223]: There's no need for prefix as the field name is defined by customer and this flatten feature is also controlled by customer.

File: src/promptflow/promptflow/_sdk/_serving/extension/otel_exporter_provider_factory.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+        except ImportError:
+            return None
+
+
+class OTelExporterProviderFactory:
+    """Factory to create OTel trace and metric exporters based on extension type."""
+
+    @staticmethod
+    def get_trace_exporters(logger, extension: ExtensionType, **kwargs):
+        trace_providers = [AppInsightTraceExporterProvider(logger), MdcTraceExporterProvider(logger)]
```
[the remark this answers is not in the document]
Comment (wxpjimmy) [reply src/promptflow/promptflow/_sdk/_serving/extension/otel_exporter_provider_factory.py:88]: this function will only be called once, so there's no need to move it out of this method.

File: src/promptflow/promptflow/_sdk/_serving/utils.py (hunk 4 above)
Comment (wxpjimmy) [reply src/promptflow/promptflow/_sdk/_serving/utils.py:147]: Yes, this is using otel sdk,  otel doesn't provide existing util,  you need to write your code to extract the context based just like the sample code you got.

File: src/promptflow/promptflow/tracing/contracts/trace.py (hunk 2 above)
Comment (wxpjimmy) [reply src/promptflow/promptflow/tracing/contracts/trace.py:19]: This is optional for serving scenario, I can remove this if there's concern.

File: src/promptflow/promptflow/_sdk/_serving/extension/extension_type.py (hunk 3 above)
Comment (wxpjimmy) [reply src/promptflow/promptflow/_sdk/_serving/extension/extension_type.py]: good suggestion, updated

File: src/promptflow/promptflow/_sdk/_serving/extension/otel_exporter_provider_factory.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    """
+    Try to get application insight connection string from environment variable.
+    app insight base exporter support these environment variables:
+    - "APPINSIGHTS_INSTRUMENTATIONKEY"
+    - "APPLICATIONINSIGHTS_CONNECTION_STRING"
+    """
+    instrumentation_key = os.getenv("AML_APP_INSIGHTS_KEY") or os.getenv("APPINSIGHTS_INSTRUMENTATIONKEY")
+    if instrumentation_key:
+        return f"InstrumentationKey={instrumentation_key}"
+    connection_str = os.getenv("APPLICATIONINSIGHTS_CONNECTION_STRING")
```
Comment (brynn-code) [inline src/promptflow/promptflow/_sdk/_serving/extension/otel_exporter_provider_factory.py:119]: Will the two env var be exposed to users? If so, we may need to document them.

Comment (wxpjimmy) [reply src/promptflow/promptflow/_sdk/_serving/extension/otel_exporter_provider_factory.py:119]: Yes, these two are what ApplicationInsightsExporter defined, we don't introduce new envs

File: src/promptflow/promptflow/tracing/contracts/trace.py (hunk 2 above)
Comment (wxpjimmy) [reply src/promptflow/promptflow/tracing/contracts/trace.py:19]: Let me remove it first,  we can add it back if there's requirement for this.

File: src/promptflow/promptflow/_sdk/_serving/app.py [hunk 10]
```diff
@@ -211,6 +204,36 @@ def version():
...
+                should_flatten = request.args.get('flatten', 'false').lower() == 'true'
+                if should_flatten:
+                    try:
+                        # try flatten the data to avoid data too big issue (especially for app insights scenario)
+                        data = json.loads(data)
+                        for k in data:
+                            span.set_attribute(k, serialize_attribute_value(data[k]))
+                    except Exception as e:
+                        logger.warning(f"Failed to flatten the feedback, fall back to non-flattern mode. Error: {e}.")
+                        span.set_attribute('feedback', data)
```
[the remark this answers is not in the document]
Comment (wxpjimmy) [reply src/promptflow/promptflow/_sdk/_serving/app.py]: updated

File: src/promptflow/promptflow/_sdk/_serving/utils.py (hunk 1 above)
Comment (wxpjimmy) [reply src/promptflow/promptflow/_sdk/_serving/utils.py:160]: this is user's data, we need to try our best to keep it as is.

**Review by zzn2 — APPROVED** [review]

**Review by brynn-code — APPROVED** [review]

**Pull request merged**